File: control/AmScopeCamera.cpp
#include "config.hpp"

#include "AmScopeCamera.hpp"

#if defined(DEBUG_MODE_CAMERA)
    #define DEBUG(...) DEBUG_OUTPUT(__VA_ARGS__)
#else
    #define DEBUG(...)
#endif

void (*AmScopeCamera::singleCaptureCallback)(uint8_t *imgData, int width, int height, ImageFormat format);
void (*AmScopeCamera::streamCaptureCallback)(uint8_t *imgData, int width, int height, ImageFormat format);

const int AmScopeCamera::liveIndex = 1;
const int AmScopeCamera::stillIndex = 1;

HAmcam AmScopeCamera::cameraHandle;

int AmScopeCamera::liveWidth;
int AmScopeCamera::liveHeight;
void *AmScopeCamera::liveData;

int AmScopeCamera::stillWidth;
int AmScopeCamera::stillHeight;
void *AmScopeCamera::stillData;

bool AmScopeCamera::liveImageGood;
bool AmScopeCamera::stillImageGood;

AmScopeCamera::AmScopeCamera() {
    singleCaptureCallback = nullptr;
    streamCaptureCallback = nullptr;
    closeCamera(); //reset to known state
}

AmScopeCamera::~AmScopeCamera() {
    closeCamera(); //reset to known state
}

bool AmScopeCamera::liveImageReady() {
    return liveImageGood;
}

bool AmScopeCamera::stillImageReady() {
    return stillImageGood;
}

void AmScopeCamera::setSingleCaptureCallback(void (*cb)(uint8_t *data, int w, int h, ImageFormat fmt)) {
    singleCaptureCallback = cb;
}

//set to null if no callback desired
void AmScopeCamera::setStreamCaptureCallback(void (*cb)(uint8_t *data, int w, int h, ImageFormat fmt)) {
    streamCaptureCallback = cb;
}

bool AmScopeCamera::isOpen() {
    return cameraHandle != NULL;
}

bool AmScopeCamera::openCamera() {
    HRESULT hr;

    if(isOpen()) {
        DEBUG("[CameraModule] Camera is already open\n");
        return true;
    }

    cameraHandle = Amcam_Open(NULL);

    if(cameraHandle == NULL) {
        DEBUG("[CameraModule] No camera found or open failed\n");
        return false;
    }

    hr = Amcam_put_eSize(cameraHandle, stillIndex);
    if(SUCCEEDED(hr))
        hr = Amcam_get_Size(cameraHandle, &stillWidth, &stillHeight);

    if (FAILED(hr)) {
        DEBUG("[CameraModule] Failed to set or get still size; hr = %d\n", hr);
        return false;
    }
    else {
        stillData = static_cast<void *>(new uint8_t[TDIBWIDTHBYTES(24 * stillWidth) * stillHeight]);
        if (NULL == stillData) {
            DEBUG("[CameraModule] Failed to allocate memory for still image data\n");
            return false;
        }
    }

    hr = Amcam_put_eSize(cameraHandle, liveIndex);
    if(SUCCEEDED(hr))
        hr = Amcam_get_Size(cameraHandle, &liveWidth, &liveHeight);

    if (FAILED(hr)) {
        DEBUG("[CameraModule] Failed to set or get live size; hr = %d\n", hr);
        return false;
    }
    else {
        liveData = static_cast<void *>(new uint8_t[TDIBWIDTHBYTES(24 * liveWidth) * liveHeight]);
        if (NULL == liveData) {
            DEBUG("[CameraModule] Failed to allocate memory for live image data\n");
            return false;
        }
    }

    DEBUG("[CameraModule] Camera opened: live resolution %dx%d, still resolution %dx%d\n", \
            liveWidth, liveHeight, stillWidth, stillHeight);
    return true;
}

void AmScopeCamera::closeCamera() {
    if(isOpen())
        Amcam_Close(cameraHandle);

    if(liveData)
        delete[] static_cast<uint8_t *>(liveData);

    if(stillData)
        delete[] static_cast<uint8_t *>(stillData);
    
    stillData = NULL;
    liveData = NULL;
    cameraHandle = NULL;
    liveImageGood = false;
    stillImageGood = false;

    DEBUG("[CameraModule] Camera closed\n");
}

bool AmScopeCamera::singleCapture() {
    HRESULT hr = Amcam_StartPullModeWithCallback(cameraHandle, amscopeCallback, NULL);
    hr = Amcam_Snap(cameraHandle, stillIndex);

    stillImageGood = false;

    if (FAILED(hr)) {
        DEBUG("[CameraModule] Failed to capture still image, hr = %d\n", hr);
        return false;
    }
    else {
        DEBUG("[CameraModule] Waiting for still image...\n");
        return true;
    }
}

bool AmScopeCamera::streamCapture() {
    HRESULT hr = Amcam_StartPullModeWithCallback(cameraHandle, amscopeCallback, NULL);

    liveImageGood = false;

    if (FAILED(hr)) {
        DEBUG("[CameraModule] Failed to capture live image, hr = %d\n", hr);
        return false;
    }
    else {
        DEBUG("[CameraModule] Waiting for live image...\n");
        return true;
    }
}

void AmScopeCamera::amscopeCallback(unsigned nEvent, void *pCallbackCtx) {
    HRESULT hr;
    AmcamFrameInfoV2 info = {};

    if (nEvent == AMCAM_EVENT_IMAGE) {
        liveImageGood = false;
        hr = Amcam_PullImageV2(cameraHandle, liveData, 24, &info);

        if (FAILED(hr)) {
            DEBUG("[CameraModule] Failed to pull image, hr = %d\n", hr);
        }
        else {
            DEBUG("[CameraModule] Live image captured.\n");
            liveImageGood = true;

            if(streamCaptureCallback)
                streamCaptureCallback(static_cast<uint8_t *>(liveData), liveWidth, liveHeight, RGB888); //modify later
        }
    }
    else if(nEvent == AMCAM_EVENT_STILLIMAGE) {
        stillImageGood = false;
        hr = Amcam_PullStillImageV2(cameraHandle, stillData, 24, &info);

        if (FAILED(hr)) {
            DEBUG("[CameraModule] Failed to pull image, hr = %d\n", hr);
        }
        else {
            DEBUG("[CameraModule] Still image captured.\n");
            stillImageGood = true;

            if(singleCaptureCallback)
                singleCaptureCallback(static_cast<uint8_t *>(stillData), stillWidth, stillHeight, RGB888); //modify later

            int count = 0;
            for(int i = 0; i < stillWidth * stillHeight; i++) {
                count += (((uint8_t *)stillData)[i] != 0);
            }

            DEBUG("count is %d\n", count);
        }
    }
    else { //for more robust operation, add more event handlers here
        DEBUG("[CameraModule] Other callback: %d\n", nEvent);
    }
}

uint8_t *AmScopeCamera::getSingleCaptureImage(uint8_t *img, int w, int h, ImageFormat fmt) {
    if(!stillImageReady())
        return nullptr;
    copyImage(static_cast<uint8_t *>(stillData), img, stillWidth, stillHeight, w, h, fmt);
    return img;
}

uint8_t *AmScopeCamera::getStreamCaptureImage(uint8_t *img, int w, int h, ImageFormat fmt) {
    if(!liveImageReady())
        return nullptr;
    copyImage(static_cast<uint8_t *>(liveData), img, liveWidth, liveHeight, w, h, fmt);
    return img;
}

void AmScopeCamera::copyImage(uint8_t *src, uint8_t *dest, int ws, int hs, int wd, int hd, ImageFormat fmt) {
    int minWidth = (ws < wd) ? ws : wd;
    int minHeight = (hs < hd) ? hs : hd;

    DEBUG("%d, %d\n", minWidth, minHeight);

    if(1) { //normally would check for format here; this assumes 24 bits per pixel
        //copy image data
        for(int y = 0; y < minHeight; y++) {
            for(int x = 0; x < minWidth; x++) {
                dest[3*(wd*y + x)] = src[3*(ws*y + x)];
                dest[3*(wd*y + x) + 1] = src[3*(ws*y + x) + 1];
                dest[3*(wd*y + x) + 2] = src[3*(ws*y + x) + 2];
            }
        }

        //if there's extra space on the right side of the destination, set its data to 0
        for(int x = minWidth; x < wd; x++) {
            for(int y = 0; y < minHeight; y++) {
                dest[3*(wd*y + x)] = 0;
                dest[3*(wd*y + x) + 1] = 0;
                dest[3*(wd*y + x) + 2] = 0;
            }
        }

        //if there's extra space at the bottom of the destination, set its data to 0
        for(int y = minHeight; y < hd; y++) {
            for(int x = 0; x < wd; x++) {
                dest[3*(wd*y + x)] = 0;
                dest[3*(wd*y + x) + 1] = 0;
                dest[3*(wd*y + x) + 2] = 0;
            }
        }
    }
}

#undef DEBUG

File: scripts/camera/amscope/amscope_camera.py
# J. Kent Wirant
# Hacker Fab
# Amscope Camera Module

from camera.camera_module import *
import amcam
import time

class AmscopeCamera(CameraModule):

    camera = None

    liveIndex = 0
    liveData = None

    stillIndex = 0
    stillData = None

    liveImageGood = False
    stillImageGood = False

    __resolutionModes = None


    def __init__(self):
        self.singleCaptureCallback = None
        self.streamCaptureCallback = None
        self.close() # reset to known state


    def __del__(self):
        self.close() # reset to known state


    def liveImageReady(self):
        return self.liveImageGood


    def stillImageReady(self):
        return self.stillImageGood


    def isOpen(self):
        return self.camera != None


    def open(self):
        if self.isOpen():
            return True

        self.camera = amcam.Amcam.Open(None)

        if self.camera == None:
            return False

        self.setResolution(self.getAvailableResolutions()[0])
        return True


    def close(self):
        if self.isOpen():
            self.camera.Close()
        
        self.stillData = None
        self.liveData = None
        self.camera = None
        self.liveImageGood = False
        self.stillImageGood = False


    def singleCapture(self):
        self.stillImageGood = False
        self.camera.StartPullModeWithCallback(self.staticCallback, self)
        self.camera.Snap(self.stillIndex)
        return True


    def streamCapture(self):
        self.liveImageGood = False
        self.camera.StartPullModeWithCallback(self.staticCallback, self)
        return True


    @staticmethod
    def staticCallback(nEvent, ctx):
        ctx.amscopeCallback(nEvent)


    def amscopeCallback(self, nEvent):
        if (nEvent == amcam.AMCAM_EVENT_IMAGE):
            self.liveImageGood = False
            self.camera.PullImageV2(self.liveData, 24, None)
            self.liveImageGood = True

            if self.streamCaptureCallback != None:
                r = self.getResolution()
                self.streamCaptureCallback(self.liveData, r, self.ImageFormat.RGB888) # modify later

        elif nEvent == amcam.AMCAM_EVENT_STILLIMAGE:
            self.stillImageGood = False
            self.camera.PullStillImageV2(self.stillData, 24, None)
            self.stillImageGood = True

            if self.singleCaptureCallback != None:
                r = self.getResolution()
                self.singleCaptureCallback(self.stillData, r, self.ImageFormat.RGB888) # modify later

        else: # for more robust operation, add more event handlers here
            pass


    def getSingleCaptureImage(img, width, height, imageFormat):
        if not self.stillImageReady():
            return None
        r = getResolution()
        self.__copyImage(self.stillData, img, r.width, r.height, width, height, imageFormat)
        return img


    def getStreamCaptureImage(imgData, width, height, imageFormat):
        if not self.liveImageReady():
            return None
        r = getResolution()
        __copyImage(self.liveData, img, r.width, r.height, width, height, imageFormat)
        return img


    def setSingleCaptureCallback(self, callback):
        self.singleCaptureCallback = callback


    def setStreamCaptureCallback(self, callback):
        self.streamCaptureCallback = callback


    # helper function
    def __copyImage(srcData, destData, ws, hs, wd, hd, imageFormat):
        minWidth = ws if (ws < wd) else wd
        minHeight = hs if (hs < hd) else hd

        if True: # normally would check for format here; this assumes 24 bits per pixel
            # copy image data
            for y in range(0, minHeight):
                for x in range(0, minWidth):
                    destData[3*(wd*y + x)] = srcData[3*(ws*y + x)]
                    destData[3*(wd*y + x) + 1] = srcData[3*(ws*y + x) + 1]
                    destData[3*(wd*y + x) + 2] = srcData[3*(ws*y + x) + 2]

            # if there's extrself*(wd*y + x) + 1] = 0
                    destData[3*(wd*y + x) + 2] = 0

            # if there's extra space at the bottom of the destination, set its data to 0
            for y in range(minHeight, hd):
                for x in range(0, wd):
                    destData[3*(wd*y + x)] = 0
                    destData[3*(wd*y + x) + 1] = 0
                    destData[3*(wd*y + x) + 2] = 0

    
    # TODO: implement separate single and stream implementations
    def getAvailableResolutions(self, mode=None):
        if self.__resolutionModes == None:
            num_resolutions = self.camera.ResolutionNumber()
            self.__resolutionModes = [None] * num_resolutions

            for i in range(0, num_resolutions):
                self.__resolutionModes[i] = self.camera.get_Resolution(i)

        return self.__resolutionModes
    
    # TODO: implement separate single and stream implementations
    def getResolution(self, mode=None):
        return self.__resolutionModes[self.liveIndex]

    # TODO: implement separate single and stream implementations
    def setResolution(self, resolution, mode=None):
        for i in range(0, len(self.__resolutionModes)):
            if self.__resolutionModes[i] is resolution or self.__resolutionModes[i] == resolution:
                self.stillIndex = i
                self.liveIndex = i
                self.camera.put_eSize(i)
                self.stillData = bytes(resolution[0] * resolution[1])
                self.liveData = bytes(resolution[0] * resolution[1])
                self.stillImageGood = False
                self.liveImageGood = False
                return True
        return False


    # camera description functions
    def getDeviceInfo(self, parameterName):
        match parameterName:
            case 'vendor':
                return self.camera.AmcamDeviceV2.displayname
            case 'name':
                return self.camera.AmcamModelV2.name
            case other:
                return None


# test suite
if __name__ == "__main__":
    def testCallback(image, resolution, format):
        print(f"{resolution[0]} {resolution[1]} {format}")

    def testCase(testName, expectedValues, actualValues):
        global testCount
        global testsPassed
        global testPrefixString

        errorStringLimit = 100

        # convert inputs to lists if necessary
        if not isinstance(expectedValues, list):
            expectedValues = [expectedValues]
        if not isinstance(actualValues, list):
            actualValues = [actualValues]
        assert len(expectedValues) == len(actualValues)

        testCount += 1
        success = True
        errorString = "(expected, got) = ["

        for i in range(0, len(expectedValues)):
            if expectedValues[i] is not actualValues[i] or expectedValues[i] != actualValues[i]:
                success = False
                if(len(errorString) <= errorStringLimit):
                    errorString += f"({expectedValues[i]}, {actualValues[i]}); "
                if(len(errorString) > errorStringLimit):
                    errorString = errorString[0:errorStringLimit] + "..."

        if success:
            testsPassed += 1
            print(testPrefixString + "PASSED Test '" + testName + "'")
        else:
            if(len(errorString) > errorStringLimit):
                errorString = errorString + "]"
            else:
                errorString = errorString[:-2] + "]"
            print(testPrefixString + "FAILED Test '" + testName + "': " + errorString)

        return success


    testCount = 0
    testsPassed = 0
    testPrefixString = "[AmscopeCamera] "

    camera = AmscopeCamera()
    openSuccess = camera.open()
    testCase("open()", True, openSuccess)

    if not openSuccess:
        print(testPrefixString + "Further testing requires connection to Amscope Camera")
    else:
        print(testPrefixString + camera.getDeviceInfo('name'))
        print(testPrefixString + camera.getDeviceInfo('vendor'))
        
        resolutions = camera.getAvailableResolutions()
        print(testPrefixString + "Resolutions: ", end='')

        expectedResolutionMode = []
        actualResolutionMode = []
        
        for r in resolutions:
            print(str(r) + " ", end='')
            camera.setResolution(r)
            expectedResolutionMode.append(r)
            actualResolutionMode.append(camera.getResolution())
        print()
        testCase("setResolution(ResolutionMode)", expectedResolutionMode, actualResolutionMode)

        camera.setStreamCaptureCallback(testCallback)
        camera.startStreamCapture()
        time.sleep(5)

    print(testPrefixString + f"Result: {testsPassed}/{testCount} tests passed")
